src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj

```diff
@@ -0,0 +1,86 @@
+(ns puppetlabs.services.jruby-pool-manager.impl.reference-pool
+  (:require [puppetlabs.services.protocols.jruby-pool :as pool-protocol]
+            [puppetlabs.services.jruby-pool-manager.impl.jruby-agents :as jruby-agents]
+            [puppetlabs.services.jruby-pool-manager.impl.jruby-internal :as jruby-internal]
+            [puppetlabs.services.jruby-pool-manager.jruby-schemas :as jruby-schemas]
+            [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n])
+  (:import (puppetlabs.services.jruby_pool_manager.jruby_schemas ReferencePool)))
+
+(extend-type ReferencePool
+  pool-protocol/JRubyPool
+
+  (fill
+    [this]
+    (jruby-agents/add-instance this 1))
+
+  (shutdown
+    [this]
+    (let [pool (jruby-internal/get-pool this)
+          cleanup-fn (get-in this [:config :lifecycle :cleanup])
+          instance (.borrowItem pool)
+          _ (.releaseItem pool instance)]
+      (jruby-internal/insert-shutdown-poison-pill pool)
+      ;; This will block until all borrows have been returned
+      (jruby-internal/cleanup-pool-instance! instance cleanup-fn)))
+
+  (lock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lock pool)))
+
+  (lock-with-timeout
+    [this timeout time-unit]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lockWithTimeout pool timeout time-unit)))
+
+  (unlock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.unlock pool)))
+
+  (borrow
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.borrowItem pool)))
+
+  (return
+    [this instance]
+    (when (jruby-schemas/jruby-instance? instance)
+      (let [pool (jruby-internal/get-pool this)
+            borrow-count (:borrow-count this)
+            max-borrows (get-in instance [:internal :max-borrows])
+            modify-instance-agent (jruby-agents/get-modify-instance-agent this)]
+        (.releaseItem pool instance)
+        (swap! borrow-count inc)
+        ;; If max-borrows is 0, never flush the instance
+        (when (and (pos? max-borrows)
+                   (>= @borrow-count max-borrows)
+                   ;; If the pool is already locked, there's a good chance
+                   ;; a flush is already in progress, and we shouldn't queue
+                   ;; another one. If the pool is locked for some other reason,
+                   ;; we'll flush later. `borrow-count` does not get set back to
+                   ;; 0 until a flush succeeds.
+                   ;; THIS IS RACY
```

Comment:
But still helps prevent duplicate flushes some (most?) of the time. I'd appreciate thoughts on better ways to do this.

Reply:
If you acquire the lock around this part then you won't retrieve borrow count until the flush is done and when you get it it will be near 0 and should therefore not trigger a new flush. (It's usually an error to check a lock instead of just acquiring it.) I think it would be ideal actually if this whole check occurred inside the function that the agent calls.

Reply:
If I'm understanding this right, putting the whole check inside the agent-called function is better because the agent will ensure that any other action the agent is performing (e.g. an in-progress flush) has completed before it calls the function again, ensuring that when we check the conditions, they will represent the post-flush state. Is that accurate? The docs on agents are a little hard to parse...

Reply:
Update: we should check the borrows both before and after locking, and if they're back to 0 after we've acquired the lock, we should skip flushing and just unlock immediately.